[Merged by Bors] - feat(data/polynomial/monic): monic.not_zero_divisor_iff

pechersky opened the pull request:
We prove that a monic polynomial is not a zero divisor. A helper lemma is proven that the product of a monic polynomial is of lesser degree iff it is nontrivial and the multiplicand is zero.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text] 
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

- [x] depends on: #8474

pechersky committed:
feat(data/polynomial/monic): monic.not_zero_divisor_iff

jcommelin commented on src/data/polynomial/monic.lean:
```diff
@@ -266,5 +266,47 @@ lemma ne_zero_of_monic (h : monic p) : p ≠ 0 :=
+
+variables [semiring R] {p : polynomial R}
+
+lemma monic.mul_lt_degree_iff_ne_one_and_mul_eq_zero (h : monic p) {q : polynomial R} :
```
```suggestion
lemma monic.mul_nat_degree_lt_iff (h : monic p) {q : polynomial R} :
```

jcommelin commented on src/data/polynomial/monic.lean:
```diff
@@ -266,5 +266,47 @@ lemma ne_zero_of_monic (h : monic p) : p ≠ 0 :=
+    simpa [hq, mul_zero, nat_degree_zero] using lt_of_le_of_ne (nat.zero_le _) hp }
+end
+
+lemma monic.not_zero_divisor_iff (h : monic p) {q : polynomial R} :
```
```suggestion
lemma monic.mul_eq_zero_iff (h : monic p) {q : polynomial R} :
```

eric-wieser replied:
This looks to me like it would benefit from left and right versions depending on which variable is Monic.

jcommelin commented on src/data/polynomial/monic.lean:
```diff
@@ -266,5 +266,47 @@ lemma ne_zero_of_monic (h : monic p) : p ≠ 0 :=
+  { rw [hq],
+    simp }
+end
+
```
I can imagine that a lemma `monic.mul_ne_zero` (of_ne_zero) might come in handy in the future. What do you think?

pechersky committed:
feat(ordered_monoid): add_eq_bot_iff

bot addition is saturating on the bottom. On the way, typeclass arguments
were relaxed to just `[add_monoid α]`, and helper simp lemmas
added for `bot`.

pechersky committed:
rename to monic.degree_mul

pechersky committed:
mul_left_eq_zero_iff

pechersky committed:
fix sorries

pechersky committed:
add is_unit cancel lemmas

pechersky committed:
Merge branch 'master' into pechersky/monic-non-zero-divisor-iff

pechersky committed:
fix referenced lemma name

pechersky committed:
add suggested TODOs
